lib/quicrq.c

```diff
@@ -1788,6 +1789,83 @@ int quicrq_receive_stream_data(quicrq_stream_ctx_t* stream_ctx, uint8_t* bytes,
     return ret;
 }
 
+/*
+ * todo: add documentation
+ */
+int quicrq_receive_warp_or_rush_stream_data(quicrq_uni_stream_ctx_t* stream_ctx, uint8_t* bytes, size_t length, int is_fin) {
+    // warp-header:grp-id [obj1 , obj2 ]
+    int ret = 0;
+
+    while (ret == 0 && length > 0) {
+        /* There may be a set of messages back to back, and all have to be received. */
+        if (stream_ctx->receive_state == quicrq_receive_done) {
+            /* Protocol violation, was not expecting any message */
+            ret = -1;
+            break;
+        }
+        else {
+            /* Receive the next message on the stream, if any */
+            int is_finished = 0;
+            uint8_t* next_bytes = quicrq_msg_buffer_store(bytes, length, &stream_ctx->message_receive, &is_finished);
+            if (next_bytes == NULL) {
+                /* Something went wrong */
+                ret = -1;
+            }
+            else
+            {
+                length = (bytes + length) - next_bytes;
+                bytes = next_bytes;
+                if (is_finished) {
+                    /* Decode the incoming message */
+                    quicrq_message_t incoming = { 0 };
+                    const uint8_t* r_bytes = quicrq_msg_decode(stream_ctx->message_receive.buffer, stream_ctx->message_receive.buffer + stream_ctx->message_receive.message_size, &incoming);
+
+                    if (r_bytes == NULL) {
+                        /* Message was incorrect */
+                        ret = -1;
+                    } else {
+                        switch (incoming.message_type) {
+                            case QUICRQ_ACTION_WARP_HEADER:
+                                stream_ctx->media_id = incoming.media_id;
+                                /* Open the media -- TODO, variants with different actions. */
+                                quicrq_log_message(stream_ctx->cnx_ctx, "UniStream %" PRIu64 ", received warp header message id= %" PRIu64,
+                                                   ", ControlStream id= %" PRIu64 " media id=%" PRIu64,
+                                                   stream_ctx->stream_id, stream_ctx->cnx_ctx->control_stream->stream_id, incoming.media_id);
+
```

Comment:
How do i find the data structure to do the lookup of control stream from the uni_stream_context

Reply:
per discussion:
extract the media_id from the warp_header and loop through the control stream from the connection_context

Reply:
You should open a block (curly braces) to cover the code from `case QUICRQ_ACTION_WARP_HEADER:` to `break;`, so that declarations like `ctrl_stream_ctx` will be scoped inside that block. Also, old C style, it is best if declarations are the first statement in the block. Line 1841 is dup from 1843. Line 1840 could be move just before 1843.